### PR TITLE
11.7.1 features merge

### PR DESCRIPTION
**Pull request by orenbm**
### What does this PR do?
Merge all the features from 11.7.
Google cloud function support in GCP authenticator

### What ticket does this PR close?
Resolves #[relevant GitHub issues](https://github.com/cyberark/conjur/issues/1813)

### CODE REVIEW

File: app/domain/authentication/authn_gcp/decoded_token.rb [hunk 1]
```diff
@@ -0,0 +1,77 @@
+# frozen_string_literal: true
```
Comment (micahlee) [inline app/domain/authentication/authn_gcp/decoded_token.rb]: How are these 3 new files fixing merge conflicts? Also, shouldn't these be integrated into the functional commits in which the conflict was introduced?

Comment (orenbm) [reply app/domain/authentication/authn_gcp/decoded_token.rb]: They don't. I rearranged the commits now so that we don't have this commit anymore

**Review by micahlee — COMMENTED** [review]
@eranha, thanks for the review request!

This is a massive PR. Is this the first time these changes have been reviewed? Or have these previously been reviewed before, and this just brings a larger branch into master?

Given the size and the fact that I haven't been involved in the GCP authenticator project, I don't feel I can approve this PR. Only raise attention to higher-level issues (as I did regarding the commit history originally).

The only issue I have with this PR at the moment is the number of Code climate violations that are reported. If these changes *were* previously reviewed, it is concerning that they were approved before. In any case, these need to be addressed either by fixing them or by ignoring them with a comment that explains why it isn't an issue.

Thanks!

File: cucumber/authenticators_gcp/features/support/authn_gcp_helper.rb [hunk 2]
```diff
@@ -75,55 +84,106 @@ def no_kid_self_signed_token
...
+
+def gce_token_non_existing_account
+  @gce_token_non_existing_account ||= read_token_file("gce_token_non_existing_account")
+end
+
+def gce_token_user_audience
+  @gce_token_user_audience ||= read_token_file("gce_token_user")
+end
+
+def gcf_identity_access_token(token_type)
```
[the remark this answers is not in the document]
Comment (eranha) [reply cucumber/authenticators_gcp/features/support/authn_gcp_helper.rb:137]: These are two different types of tokens.

File: cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb [hunk 3]
```diff
@@ -24,25 +24,76 @@
...
-  gcp_identity_access_token(
+Given(/^I obtain an? (valid|standard_format|user_audience|invalid_audience|non_existing_host|non_rooted_host|non_existing_account) GCE identity token$/) do |token_type|
+  gce_identity_access_token(
     token_type.to_sym
   )
 end
 
 # Authenticates with Conjur GCP authenticator
-Given(/I authenticate (?:(\d+) times? in (\d+) threads? )?with authn-gcp using (no|empty|self signed|no kid|obtained|valid) token and (non-existing|existing) account/) do |num_requests, num_threads, token_state, account|
+Given(/I authenticate (?:(\d+) times? in (\d+) threads? )?with authn-gcp using (no|empty|self signed|no kid|obtained GCE|valid GCE) token and (non-existing|existing) account/) do |num_requests, num_threads, token_state, account|
```
[the remark this answers is not in the document]
Comment (eranha) [reply cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb:97]: Cucumber does not allow breaking.

File: cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb [hunk 4]
```diff
@@ -1,19 +1,19 @@
 # Enable selective setting of GCP annotations
-Given(/^I set (invalid )?"authn-gcp\/(service-account-id|service-account-email|project-id|instance-name|invalid-key)" annotation to host "([^"]*)"$/) do |invalid, annotation_name, hostname|
+Given(/^I set (invalid )?"authn-gcp\/(service-account-id|service-account-email|project-id|instance-name|invalid-key)" GCE annotation to host "([^"]*)"$/) do |invalid, annotation_name, hostname|
```
[the remark this answers is not in the document]
Comment (eranha) [reply cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb:2]: Not used in any of the other step files.

File: cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb [hunk 5]
```diff
@@ -24,25 +24,76 @@
   set_annotation_to_resource("authn-gcp/#{annotation_name}", annotation_value)
 end
 
-# Sets all GCP annotations
-Given(/^I set all valid GCP annotations to (user|host) "([^"]*)"$/) do |role_type, hostname|
+Given(/^I set "authn-gcp\/(service-account-id|service-account-email|project-id|instance-name)" GCF annotation to host "([^"]*)"$/) do |annotation_name, hostname|
```
[the remark this answers is not in the document]
Comment (eranha) [reply cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb:27]: Not used in any of the other step files.

File: cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb [hunk 6]
```diff
@@ -24,25 +24,76 @@
...
+end
+
+Given(/^I set all valid GCF annotations to (user|host) "([^"]*)"$/) do |role_type, hostname|
+  i_have_a_resource role_type, hostname
+
+  set_annotation_to_resource("authn-gcp/service-account-id", gcf_service_account_id)
+  set_annotation_to_resource("authn-gcp/service-account-email", gcf_service_account_email)
+end
+
+Given(/^I set "authn-gcp\/(service-account-id|service-account-email)" GCF annotations to (user|host) "([^"]*)"$/) do |annotation_type, role_type, hostname|
```
[the remark this answers is not in the document]
Comment (eranha) [reply cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb:67]: Not used in any of the other step files.

File: cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb [hunk 7]
```diff
@@ -24,25 +24,76 @@
...
+  when "instance-name"
+    annotation_value = "_"
+  else
+    raise "Incorrect annotation name: '#{annotation_name}', expected: service-account-id|service-account-email|project-id|instance-name"
+  end
+
+  set_annotation_to_resource("authn-gcp/#{annotation_name}", annotation_value)
+end
+
+Given(/^I set "authn-gcp\/(service-account-id|service-account-email)" annotation with value: "([^"]*)" to host "([^"]*)"$/) do |annotation_name, annotation_value, hostname|
```
[the remark this answers is not in the document]
Comment (eranha) [reply cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb:46]: Not used in any of the other step files.

File: app/domain/authentication/authn_gcp/validate_resource_restrictions_match_jwt.rb [hunk 8]
```diff
@@ -38,14 +38,26 @@ def resource_type_without_prefix(resource_type)
...
         end
+
+        validate_resource_exists_in_token(resource_type, resource_value)
+
+        resource_value
+      end
+
+      def validate_resource_exists_in_token(resource_type, resource_value)
+        if resource_value.nil? || resource_value.empty?
+          raise Errors::Authentication::AuthnGcp::ResourceRestrictionNotFoundOrEmpty.new(resource_type)
```
[the remark this answers is not in the document]
Comment (eranha) [reply app/domain/authentication/authn_gcp/validate_resource_restrictions_match_jwt.rb:57]: This pattern is acceptable.

File: cucumber/authenticators_gcp/features/support/authn_gcp_helper.rb [hunk 9]
```diff
@@ -75,55 +84,106 @@ def no_kid_self_signed_token
   JWT.encode exp_payload, rsa_private, 'RS256'
 end
 
-def gcp_identity_access_token(token_type)
+def gce_identity_access_token(token_type)
```
[the remark this answers is not in the document]
Comment (eranha) [reply cucumber/authenticators_gcp/features/support/authn_gcp_helper.rb:86]: The code does what it is intended to do, ignored.

File: cucumber/authenticators_gcp/features/support/authn_gcp_helper.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (eranha) [reply cucumber/authenticators_gcp/features/support/authn_gcp_helper.rb:137]: The code does what it is intended to do, ignored.

File: cucumber/api/features/support/authz_helpers.rb [hunk 10]
```diff
@@ -39,6 +39,10 @@ def set_annotation_to_resource(name, value, resource = @current_resource)
       resource.add_annotation name: name, value: value
     end
   end
+
+  def remove_resource_all_annotations(resource = @current_resource)
+    resource.annotations.each { |a| a.delete }
```
[the remark this answers is not in the document]
Comment (eranha) [reply cucumber/api/features/support/authz_helpers.rb:44]: Changed as suggested.

File: cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (eranha) [reply cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb:67]: Not used anywhere, ignored.

File: cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (eranha) [reply cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb:46]: Not used anywhere, ignored.

File: cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (eranha) [reply cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb:27]: Not used anywhere, ignored.

File: cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (eranha) [reply cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb:2]: Not used anywhere, ignored.

File: cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (eranha) [reply cucumber/authenticators_gcp/features/step_definitions/authn_gcp_steps.rb:97]: Cucumber does not allow a line break here, ignored.

File: cucumber/authenticators_gcp/features/support/authn_gcp_helper.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (eranha) [reply cucumber/authenticators_gcp/features/support/authn_gcp_helper.rb:137]: By design, ignored.

File: cucumber/authenticators_gcp/features/support/authn_gcp_helper.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (eranha) [reply cucumber/authenticators_gcp/features/support/authn_gcp_helper.rb:86]: By design, ignored

File: app/domain/errors.rb [hunk 11]
```diff
@@ -363,16 +363,24 @@ module AuthnGcp
...
       )
 
-      TokenClaimNotFoundOrEmpty = ::Util::TrackableErrorClass.new(
-        msg:  "Claim '{0-claim-name}' not found or empty in token. " \
-              "Verify that you requested the token using 'format=full'",
+      ResourceRestrictionNotFoundOrEmpty = ::Util::TrackableErrorClass.new(
+        msg:  "Resource restriction '{0-resource-restriction-name}' does not exists " \
+            "resource in JWT token. " \
+            "Verify that you configured the host with permitted restrictions. " \
+            "In case of Compute Engine token verify that you requested the token using 'format=full'" ,
```
[the remark this answers is not in the document]
Comment (eranha) [reply app/domain/errors.rb:370]: Fixed.

File: app/domain/authentication/authn_gcp/decoded_token.rb [hunk 12]
```diff
@@ -17,30 +17,60 @@ def initialize(decoded_token_hash:, logger:)
...
 
-        unless token_claim_value
-          raise Errors::Authentication::AuthnGcp::TokenClaimNotFoundOrEmpty, token_claim
+        required_token_claim_value
+      end
+
+      def optional_token_claim_value(optional_token_claim)
+        optional_token_claim_value = token_claim_value(optional_token_claim)
+
+        if optional_token_claim_value.nil? || optional_token_claim_value.empty?
```
[the remark this answers is not in the document]
Comment (eranha) [reply app/domain/authentication/authn_gcp/decoded_token.rb:52]: Should check for nil, ignored.

File: app/domain/authentication/authn_gcp/update_authenticator_input.rb [hunk 13]
```diff
@@ -55,9 +63,14 @@ def decoded_credentials
 
       def validate_audience
         if audience_parts.length != 3 ||
-            audience_parts[0] != "conjur" ||
-            audience_parts[1] != account
+            audience_parts[0] != "conjur"
           raise Errors::Authentication::AuthnGcp::InvalidAudience, audience
+        elsif audience_parts[1] != account
```
[the remark this answers is not in the document]
Comment (eranha) [reply app/domain/authentication/authn_gcp/update_authenticator_input.rb:68]: Should check for nil, ignored.

File: app/domain/authentication/authn_gcp/validate_resource_restrictions_match_jwt.rb [hunk 14]
```diff
@@ -38,14 +38,26 @@ def resource_type_without_prefix(resource_type)
...
+          resource_value = @decoded_token.service_account_email
         end
+
+        validate_resource_exists_in_token(resource_type, resource_value)
+
+        resource_value
+      end
+
+      def validate_resource_exists_in_token(resource_type, resource_value)
+        if resource_value.nil? || resource_value.empty?
```
[the remark this answers is not in the document]
Comment (eranha) [reply app/domain/authentication/authn_gcp/validate_resource_restrictions_match_jwt.rb:56]: By design, ignored.

File: app/domain/authentication/authn_gcp/decoded_token.rb [hunk 15]
```diff
@@ -17,30 +17,60 @@ def initialize(decoded_token_hash:, logger:)
...
+      def initialize_optional_claims
+        @service_account_email = optional_token_claim_value(EMAIL_TOKEN_CLAIM_NAME)
+        @project_id = optional_token_claim_value(PROJECT_ID_TOKEN_CLAIM_NAME)
+        @instance_name = optional_token_claim_value(INSTANCE_NAME_TOKEN_CLAIM_NAME)
+      end
+
+      def required_token_claim_value(required_token_claim)
+        required_token_claim_value = token_claim_value(required_token_claim)
+
+        if required_token_claim_value.nil? || required_token_claim_value.empty?
```
[the remark this answers is not in the document]
Comment (eranha) [reply app/domain/authentication/authn_gcp/decoded_token.rb:40]: Should check for nil, ignored.

File: app/domain/authentication/authn_gcp/validate_resource_restrictions_match_jwt.rb (hunk 14 above)
[the remark this answers is not in the document]
Comment (eranha) [reply app/domain/authentication/authn_gcp/validate_resource_restrictions_match_jwt.rb:56]: Should check for nil, ignored.

**Review by nessiLahav — APPROVED** [review]

**Pull request merged**